### PR TITLE
Implement lampager-cakephp

### PR DESCRIPTION
**Pull request by chitoku-k**
🎉Happy New Year 🎉 

It's lovely to release a brand new product from Lampager team! This PR contains the solid implementation of the CakePHP adaptor of lampager that is tested with the both MySQL and SQLite. Most of the functionalities are covered with the tests currently written with some remaining part whining uncovered seen below.

## Features
- CakePHP integration
  - [x] for `Table` via `Lampager\Model\Behavior\LampagerBahavior`
  - [x] for `Controller` via `Lampager\Datasource\Paginator`
- Tests
  - [ ] Fully-covered unit tests
  - [x] Mostly-covered unit tests with the unexceptional scenarios
- Database Support
  - [x] MySQL
  - [x] SQLite (see note below)

## Coverage

image

## Note

A special `SqliteCompiler` is bundled in order to let it successfully compile `UNION ALL` with ordered `SELECT` statements because the query builder of the CakePHP 3.x does not handle that case for SQLite.

### CODE REVIEW

File: src/ORM/Query.php [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+     * Create query based on the existing query. This factory copies the internal
+     * state of the given query to a new instance.
+     *
+     * @param BaseQuery $query
+     */
+    public static function fromQuery($query)
+    {
+        $obj = new static($query->getConnection(), $query->getRepository());
+
+        foreach (get_object_vars($query) as $k => $v) {
```
Comment (mpyw) [inline src/ORM/Query.php]: How about `clone` operator? It just behaves shallow copying.

File: src/ORM/Query.php [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                 * @var string $key       The column
+                 * @var string $condition The order
+                 */
+                $this->_paginator->orderBy($key, $condition);
+            }
+
+            if ($condition instanceof OrderClauseExpression) {
+                $generator->resetCount();
+
+                if (!preg_match('/ (?<direction>ASC|DESC)$/', $condition->sql($generator), $matches)) {
```
Comment (mpyw) [inline src/ORM/Query.php]: I'm not familiar with CakePHP 3, but case-insensitive matching looks better.

File: src/ORM/Query.php (hunk 1 above)
Comment (chitoku-k) [reply src/ORM/Query.php]: This method copies all the properties from the given `\Cake\ORM\Query` to a `\Lampager\Cake\ORM\Query` that extends it. How can it be written with `clone` operator while they differ in their types?

File: src/ORM/Query.php (hunk 2 above)
Comment (chitoku-k) [reply src/ORM/Query.php]: The internal implementation of the `\Cake\Database\Expression\OrderClauseExpression` already normalizes the direction to uppercase.
https://github.com/cakephp/cakephp/blob/3.8.8/src/Database/Expression/OrderClauseExpression.php#L43

File: src/ORM/Query.php (hunk 1 above)
Comment (mpyw) [reply src/ORM/Query.php]: That was my misunderstanding, okay

**Comment by chitoku-k** [comment]
@mpyw Thanks for reviewing! Merging this now and adding more tests to cover the remaining part later on.

**Pull request merged**